### ISSUE TITLE
StackOverflowError on circular project dependencies

### ISSUE THREAD

**Issue by echebbi** [cucumber/cucumber-eclipse#360]
## Expected behavior

A project with the _Cucumber_ nature builds without error.

## Current behavior

A project with the _Cucumber_ nature produces `StackOverflowError` during the build.

## Context

I am using Cucumber to test Eclipse plug-ins and I got a `StackOverflowError` each time my project (which has the Cucumber nature) is built.

The stacktrace is the following:
```java
java.lang.StackOverflowError
	at org.eclipse.core.internal.dtree.DeltaDataTree.lookup(DeltaDataTree.java:685)
	at org.eclipse.core.internal.watson.ElementTree.includes(ElementTree.java:533)
	at org.eclipse.core.internal.resources.Workspace.getResourceInfo(Workspace.java:1752)
	at org.eclipse.core.internal.resources.Resource.getResourceInfo(Resource.java:1166)
	at org.eclipse.core.internal.resources.Project.getReferencedProjects(Project.java:467)
	at cucumber.eclipse.editor.steps.UniversalStepDefinitionsProvider.getStepDefinitions(UniversalStepDefinitionsProvider.java:45)
	at cucumber.eclipse.editor.steps.UniversalStepDefinitionsProvider.getStepDefinitions(UniversalStepDefinitionsProvider.java:47)
	at cucumber.eclipse.editor.steps.UniversalStepDefinitionsProvider.getStepDefinitions(UniversalStepDefinitionsProvider.java:47)
        ...
```

## Origin of the issue

I didn't have the time to check exactly what happens but it seems that referenced projects form circular dependencies, which causes an infinite loop. This is likely caused by fragments or optional dependencies, which are perfectly handled at runtime.

The incriminated method is [UniversalStepDefinitionsProvider#getStepDefinitions](https://github.com/cucumber/cucumber-eclipse/blob/40f6a4b9fe2a9ee3677955b140c016cba4798d3c/cucumber.eclipse.editor/src/main/java/cucumber/eclipse/editor/steps/UniversalStepDefinitionsProvider.java#L41). I believe that it can easily be fixed with a cache to prevent analyzing the same project twice (which could even improve performances).

I may be able to propose a PR during the incoming week.

**Comment by laeubi** [comment]
What do you mean by "circular project dependencies" AFAIK Eclipse does not support circles in project dependencies.

**Comment by echebbi** [comment]
I actually discovered that calling `project.getReferencedProjects()` on a project owning the _Plug-in development_ nature will also return all the local projects that are specified as dependencies within the project's _MANIFEST.MF_ files. By "local projects" I mean the projects that are in the workspace.

I am trying things a bit tricky so there may be something wrong with my dependencies but still, this error is very inconvenient since it:

- produces a StackOverflowError each time the project is built,
- forces the IDE to be restarted.

In case you're interested in how exactly the error occurs I can try to reproduce it in a simple case and share it.

**Comment by echebbi** [comment]
Fixed by #363.

**Comment by namrathaPhenoTips** [comment]
Hi there, while I am running my cucumber feature files I am getting these errors : java.lang.StackOverflowError
	at org.openqa.selenium.By$ByXPath.findElement(By.java:279)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:314)
	at org.openqa.selenium.By$BaseW3CLocator.findElement(By.java:399) 
	
The tests runs file with Junit Runner and through Maven, but running a cucumber feature file shows these errors, can you please help me
